Adding aggregations support for the `_ignored` field

eyalkoren opened the pull request:
Closes #59946

The actual implementation needs to take into account how we decide to implement #101153.
One of the [original requirements](https://github.com/elastic/elasticsearch/issues/59946#issuecomment-674171759) was to stop making the `_ignored` field stored while adding `doc_values` to it. However, the current proposal of #101153 assumes that we can access the original order of the field's content, which means we must keep it stored as well. The eventual approach will determine the implementation of this change.

eyalkoren committed:
Adding aggregations support for the _ignored field

eyalkoren committed:
Update docs/changelog/101373.yaml

elasticsearchmachine commented:
Hi @eyalkoren, I've created a changelog YAML for you.

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/IgnoredFieldMapper.java:
```diff
@@ -59,15 +80,34 @@ public Query existsQuery(SearchExecutionContext context) {
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(FieldDataContext fieldDataContext) {
+            if (hasDocValues() == false) {
+                throw new IllegalArgumentException(
```
What happens when you're aggregating over a data stream that contains both old and new indices? Will this return a partial failure? How does Kibana/Lens handle the failure?

eyalkoren replied:
I don't know. I will look for a way to add an integration test for that

eyalkoren replied:
Since it's a general question about aggregations on a field in a data stream, I guess we can add a test that is not related to the `_ignored` field. 
Something like:
1. create a component template that has a mapping for field `foo` with `"doc_values": false`
2. create a data stream index template that uses this mapping
3. index a document
5. update the component template so that the mapping for field `foo` is changed to `"doc_values": true`
6. rollover the data stream
7. index another document
8. check aggregation on field `foo`

Would that cover it?

javanna replied:
you would indeed get partial failures in this case, one failure for each shard belonging to the old indices, as they can't be aggregated on.

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/IgnoredFieldMapper.java:
```diff
@@ -29,16 +41,25 @@ public static class Defaults {
+        private IgnoredFieldType(boolean hasDocValues) {
+            super(NAME, true, true, hasDocValues, TextSearchInfo.SIMPLE_MATCH_ONLY, Collections.emptyMap());
         }
 
```
I wish I could  comment on parts of this file that were not modified...

You'll probably want to delete the `existsQuery` method so that it can do a faster exists query that leverages doc values.

eyalkoren replied:
Done. The experts review should validate that as well.

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/IgnoredFieldMapper.java:
```diff
@@ -59,15 +80,34 @@ public Query existsQuery(SearchExecutionContext context) {
+        public IndexFieldData.Builder fielddataBuilder(FieldDataContext fieldDataContext) {
+            if (hasDocValues() == false) {
+                throw new IllegalArgumentException(
+                    "aggregations for the '" + typeName()  + "' field are supported from version + " + AGGS_SUPPORT_VERSION
```
Maybe mention that it's about when the index has been created.

eyalkoren committed:
spotless

eyalkoren committed:
Merge remote-tracking branch 'eyalkoren/ignore_field_aggs' into ignore_field_aggs

eyalkoren committed:
Adjust tests

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into ignore_field_aggs

eyalkoren committed:
Add exists query using doc_values

eyalkoren committed:
Improve error message for unsupported aggs

eyalkoren committed:
ohh come on

eyalkoren commented:
@javanna please take a look to see if this is the right direction.
Some specific input I'd be happy to get:
- Only for the sake of #59946, we don't need to keep the `_ignored` field stored. However, for #101153 we may need it. What do you think we should do as part of this PR?
- Can you answer @felixbarny's [question](https://github.com/elastic/elasticsearch/pull/101373#discussion_r1373027894)? Can you provide an advice of how to create a test for that?

javanna commented:
> Only for the sake of https://github.com/elastic/elasticsearch/issues/59946, we don't need to keep the _ignored field stored. However, for https://github.com/elastic/elasticsearch/issues/101153 we may need it. What do you think we should do as part of this PR?

I am leaning towards removing the stored field, but I think this requires further discussion, because it is going to be very difficult to go back to relying on the order if we stop storing the field. We said we won't focus for now on the ignored reason, and I would add that we should try not to have positional logic around it when we do so. Could you make the changes necessary to retrieve the field from doc values and stop storing the field in the meantime?

> Can you provide an advice of how to create a test for that? EDIT: I proposed a test scenario, let me know whether you agree that it covers what we are looking for

I am not sure that we need to test for that. We know we are going to have partial failures in that case.

eyalkoren commented:
Thanks for the feedback @javanna 🙏 

> Could you make the changes necessary to retrieve the field from doc values and stop storing the field in the meantime?

I will give it a try 🙂 

>  I am not sure that we need to test for that. We know we are going to have partial failures in that case.

OK, then I guess the next question is - what do we do about that? Only document it as a caveat related to aggregating on existing data streams? Something else?

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into ignore_field_aggs

felixbarny commented:
> OK, then I guess the next question is - what do we do about that? Only document it as a caveat related to aggregating on existing data streams? Something else?

IMO, it would be enough to document that in the same place where you document that the field is supports aggregations now.

eyalkoren committed:
WIP: making field not stored

salvatore-campagna commented:
Having a look at this before starting to work on it and complete what is (maybe) missing.

My understanding is that in future we would like to enable synthetic source for logs too. Actually we would need to enable it sooner rather than later to avoid that introducing it later causes backward compatibility issues.

If the goal is to move logs to synthetic source we would not be able to keep the original order anyway for values of the `_ignored` field. Moreover, we would also lose duplicates (in my understanding this is not really an issue).

@felixbarny is this an issue? Why do we need to preserve ordering?

felixbarny commented:
I think it’s ok to not preserve the order.

salvatore-campagna committed:
Merge branch 'main' into ignore_field_aggs

javanna commented:
Agreed, let's remove the stored field and rely on doc_values for retrieval instead, with the known limitations.

salvatore-campagna committed:
fix: _ignored field in meta field list

salvatore-campagna committed:
fix: make _ignored field stored

Metadata fields like _ignored need to be 'stored' in order to be
able to include them in the response.

salvatore-campagna commented:
@javanna @felixbarny I am pushing a commit which makes the field 'stored' again because I would like to see if that makes the CI green.

My understanding is that metadata fields like `_ignored` need to be stored to be correctly returned in the response. If they are not they will be missing and YAML tests will fail. If the CI is green I will ivestigate further how the fect phase works when dealing with metadata fields which are not stored.

salvatore-campagna commented:
Almost there...after making the field stored (and including doc values) I see all tests green but the following one:

https://gradle-enterprise.elastic.co/s/aqxgf3itnzdvc/tests/:server:test/org.elasticsearch.index.mapper.IgnoredFieldMapperTests/testFetchIgnoredFieldValue

I need to have a look at the fetch phase and specifically how metadata fields are handled. Probably will need to move loading of the `_ignored` field from the `StoredFieldsPhase` to `FetchDocValuesPhase`.

salvatore-campagna committed:
fix: do not use a mock for SearchExecutionContext

salvatore-campagna committed:
fix: decrement reference count for response objects

javanna commented:
@salvatore-campagna you need to switch the field to b retrieved from doc_values instead of from stored fields. That is controlled by the `valueFetcher` method in the mapped field type impl. It currently uses a `StoredValueFetcher,` which you should be able to replace with a `DocValueFetcher`. We do need to handle bw comp though, because indices created before this change will need to still retrieve from stored fields.

salvatore-campagna commented:
> @salvatore-campagna you need to switch the field to b retrieved from doc_values instead of from stored fields. That is controlled by the `valueFetcher` method in the mapped field type impl. It currently uses a `StoredValueFetcher,` which you should be able to replace with a `DocValueFetcher`. We do need to handle bw comp though, because indices created before this change will need to still retrieve from stored fields.

@javanna as far as I understand we have two types in the mapper (to deal with backward compatibility), `IgnoredFieldType` and `LegacyIgnoredFieldType` the first one using a `ValueFetcher` which relies on doc values, the second one using a `ValueFetcher` relying on the field being `stored`.

See that I use a `DocValueFetcher`.

```
return new DocValueFetcher(docValueFormat(format, null), context.getForField(this, FielddataOperation.SEARCH));
```

Anyway, I don't see it used and inddeed I also rad a comment for the `valueFecther` method:

```
Create a helper class to fetch field values during the {@link FetchFieldsPhase}.
New field types must implement this method in order to support the search 'fields' option. **Except
for metadata fields**, field types should not throw {@link UnsupportedOperationException} since this
could cause a search retrieving multiple fields (like "fields": ["*"]) to fail.
```

Not sure if the comment is up-to-date but it says "Except for metadata fields. My debugging and this comment drive me into thinking that a `ValueFetcher` is not used for metadata fields...and if that is the case I am not aware of the reason (maybe more than saying "Except for metadata fields" the comment should explain WHY that is not the case for metadata fields?

javanna commented:
Got it, I believe your assessment is correct. That is because metadata fields that are stored like `_ignored` or `_source` are fetched via a separate sub fetch phase, see https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/search/fetch/subphase/StoredFieldsPhase.java . The field needs to be removed from there, but under a version conditional. Next, can you check if fetching the field explicitly using the fields parameter of the search API gets you the value through the value fetcher? I suspect we will then need to figure out how to make it appear at the same level as _source in the search response by default, but pulling from doc_values (via value fetcher) and not from stored.

salvatore-campagna commented:
Based on my investigation and after talking to @javanna fetching metadata fields is, at the moment expected to happen in `StoredFieldPhase` as a result of the field being stored, which has been the default so far. This means that making the field non-stored the field is not there when the `StoredFieldPhase` tries to load stored fields.

Anyway there are other phases loading data from doc values, such as `FetchVersionPhase` and `SeqNoPrimaryTermsPhase`, which I plan to "mimic" to be able to load the `_ignored` field from doc values.
From a performance standpoint I do not see possible performance issues as doc ids are iterated on a per-segment basis which means we do not incur the penalty of loading (and decompressing block) data from sparse Lucene segments (we always load data from all documents in a segment before moving to the next segment).

salvatore-campagna committed:
fix: include a new sub-fetch phase fetching the _ignored field from doc values

salvatore-campagna committed:
fix: index the _ignored field

salvatore-campagna committed:
fix: remove fetch phase if stored fields are disabled

salvatore-campagna committed:
fix: introduce the new phase in yaml tests

salvatore-campagna committed:
Merge branch 'main' into ignore_field_aggs

javanna commented on docs/reference/mapping/fields/ignored-field.asciidoc:
```diff
@@ -43,3 +43,20 @@ GET _search
 }
 --------------------------------------------------
+
+Since 8.12.0, the `_ignored` field supports aggregations as well.
```
The version will need updating

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/IgnoredFieldMapper.java:
```diff
@@ -48,27 +67,71 @@ public String typeName() {
-            // quality of the data, so we may use a slow query. It shouldn't
-            // be too slow in practice since the number of unique terms in this
-            // field is bounded by the number of fields in the mappings.
             return new TermRangeQuery(name(), null, null, true, true);
```
I don't know why the comment was removed, I would leave it.

felixbarny replied:
Since we have doc_values for the `_ignored` field, I was hoping this would also have a positive effect on the exists query. We're relying on an exists query on the  `_ignored` field in the new dataset quality page to show users the percentage of documents in their data streams that are degraded i.e. that have ignored fields.

javanna replied:
right but this comment is on the legacy impl of the exists query, which stays the same, so the comment still applies. The new impl does a plain FieldExistsQuery based on doc_values, and that is definitely better, hence no comment on that one.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/IgnoredFieldMapper.java:
```diff
@@ -48,27 +67,71 @@ public String typeName() {
+
+        @Override
+        public Query existsQuery(SearchExecutionContext context) {
+            return new FieldExistsQuery(name());
```
do we need to override this method? Can't we rely on the default implementation from MappedFieldType?

salvatore-campagna replied:
This is in the legacy type so I don't think I added it...

javanna replied:
I think you should be able to remove it.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/IgnoredFieldMapper.java:
```diff
@@ -48,27 +67,71 @@ public String typeName() {
-            context.doc().add(new StringField(NAME, field, Field.Store.YES));
+        MappedFieldType mappedFieldType = fieldType();
+        for (String ignoredField : context.getIgnoredFields()) {
+            if (mappedFieldType.hasDocValues()) {
```
are doc values configurable for the _ignored field, meaning can they be disabled via configuration?

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/IgnoredFieldMapper.java:
```diff
@@ -48,27 +67,71 @@ public String typeName() {
+            if (mappedFieldType.hasDocValues()) {
+                context.doc().add(new SortedSetDocValuesField(fieldType().name(), new BytesRef(ignoredField)));
+            }
+            if (mappedFieldType.isIndexed()) {
```
Same as doc_values, is the fact that we create an inverted index for the field configurable? It was not before, and I was under the impression that we should not make that configurable.

javanna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchIgnoredMetadataPhase.java:
```diff
@@ -0,0 +1,68 @@
+    @Override
+    public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) throws IOException {
+        StoredFieldsContext storedFieldsContext = fetchContext.storedFieldsContext();
+        if (storedFieldsContext == null || storedFieldsContext.fetchFields() == false) {
```
can you clarify why you need to check storedFieldContext here?

salvatore-campagna replied:
So this was a stored field before being a doc values field and this check is necessary to make sure that we do not include it when fetching store fields are disabled. If I understand correctly we have a way to disable fetching stored field and this makes sure we keep the same behaviour.

javanna replied:
ok thanks for explaining. Disabling fetching of stored fields affects metadata fields that are stored, but not `_index`. This seems to me like a strange feature. I would probably rather always return metadata fields. With that said, I would adapt the test in `SearchHitsTests` instead, and accept that `stored: _none_` will no longer affect `_ignored` when it is no longer stored. That seems consistent with the current behaviour, although it leaks the implementation details, but that is kind of the pitfall of `stored: _none_` anyways. I think it would be worse to pretend that `_ignored` is still stored for the sake of not changing behaviour.

salvatore-campagna replied:
This way, anyway, users might not control whether the _ignored field is returned or not anymore right? Considering that is an array of strings might, at least, affect a bit serialisation size of the response for users which are not interested and potentially affect latency when fetching many documents?

javanna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchIgnoredMetadataPhase.java:
```diff
@@ -0,0 +1,68 @@
+
+            @Override
+            public void setNextReader(LeafReaderContext readerContext) throws IOException {
+                ignoredFields = readerContext.reader().getSortedSetDocValues(IgnoredFieldMapper.NAME);
```
I wonder if we should make the retrieval go through value fetcher instead of directly to doc_values. One possibility is to add retrieval of metadata fields to the existing FetchFieldsPhase. This could help also delegating the version based logic and centralizing it, because there is already one or the other field type available based on the index created version. Here is my proposal, to summarize:

- remove retrieval of the stored field from StoredFieldPhase, unconditionally
- add retrieval of the _ignored meta field to FetchFieldsPhase (the phase will have to run unconditionally then and set meta fields and not only document fields)
- make the retrieval go through value fetcher
- version based logic in a single place (the existing getInstance)

Does this make sense? Am I missing anything?

salvatore-campagna replied:
Do you mean the retrieval of all metafields?

javanna commented on server/src/internalClusterTest/java/org/elasticsearch/index/mapper/IgnoreFieldIT.java:
```diff
@@ -0,0 +1,144 @@
+        assertEquals(NUMERIC_FIELD_NAME, ignoredField.getValue());
+    }
+
+    public void testExistsQuery() {
```
This should probably be added as a unit test to IgnoredFieldTypeTests? Same probably for the other two tests you have? Can we also have a unit test that verifies the versioning logic?

javanna commented on server/src/internalClusterTest/java/org/elasticsearch/index/mapper/IgnoreFieldIT.java:
```diff
@@ -0,0 +1,144 @@
+import static org.hamcrest.Matchers.hasSize;
+
+@SuppressWarnings("resource")
+public class IgnoreFieldIT extends ESIntegTestCase {
```
does this need to be an integ test case? Could it be a single node test case instead? Even better, there are probably parts that could become unit test and not need to spin up a cluster with multiple nodes? Can you check the existing coverage for _ignored in the existing yaml tests? Shall we add a basic yaml test that aggregates on it? We also need to add a bw compatibility test that verifies the behaviour against a mixed cluster.

salvatore-campagna replied:
I think I need to add a few tests where I run at least a handful of aggregations on it...if I try to imagine possible use case scenarios I imagine top[Hits/Metrics] aggs maybe nested in daily/weekly/monthly histograms...date range, histogram and cardinality at least.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/IgnoredFieldMapper.java:
```diff
@@ -30,14 +43,20 @@ public static class Defaults {
+    public static final TypeParser PARSER = new FixedTypeParser(c -> getInstance(c.indexVersionCreated()));
 
-    public static final class IgnoredFieldType extends StringFieldType {
+    public static MetadataFieldMapper getInstance(IndexVersion indexVersion) {
```
does getInstance need to be public?

javanna reviewed (commented):
I had a look and left some comments

salvatore-campagna committed:
fix: include _ignored field as stored for backward compatibility

salvatore-campagna committed:
fix: a few review comments

salvatore-campagna committed:
fix: a couple more tests

salvatore-campagna committed:
fix: a few more tests

salvatore-campagna committed:
remove assertion which is version dependent

salvatore-campagna committed:
fix: update skip version

salvatore-campagna committed:
fix: _ignored field does not have terms vector anymore since it is not stored

salvatore-campagna committed:
fix: release response objects

salvatore-campagna committed:
fix: do not rely on order which is lost with doc values

salvatore-campagna committed:
fix: try to release again everything

salvatore-campagna commented on server/src/internalClusterTest/java/org/elasticsearch/action/termvectors/GetTermVectorsIT.java:
```diff
@@ -1048,9 +1047,9 @@ public void testTermVectorsWithIgnoredField() throws IOException, InterruptedExc
-        // do a tv request for all fields, _ignored should be returned
+        // do a tv request for all fields, _ignored should not be returned as a result of not being stored but just having doc values
         TermVectorsResponse resp = client().prepareTermVectors("index", "1").setSelectedFields("*").get();
-        assertThat(resp.getFields().terms("_ignored").size(), greaterThan(0L));
```
I will need to restore this test to the original and have the field type call `setStoreTermVectors(true)` to make the _ignored field store also term vectors.

salvatore-campagna committed:
test: restore test checking term vectors for the _ignored field

salvatore-campagna committed:
fix: remove loading on meta fields from StoredFieldsPhase

salvatore-campagna committed:
fix: remove FetchIgnoredMetadataPhase and use FetchFieldsFetcher for meta fields

salvatore-campagna committed:
fix: remove FetchIgnoredMetadataPhase

salvatore-campagna committed:
fix: do not include unmapped fields

salvatore-campagna committed:
fix: FetchFieldsPhase always executed to fetch metadata fields

salvatore-campagna committed:
fix: evoid creation of empty list

salvatore-campagna committed:
fix: move  metadata fields to MetadataFetcher

salvatore-campagna committed:
fix: remove unused class

salvatore-campagna committed:
fix: do not lead stored fields if not required

salvatore-campagna committed:
fix: test terms aggregation on _ignored field

salvatore-campagna committed:
fix: forgot to make the field not stored

salvatore-campagna committed:
fix: prevent NPE when storedFieldsContext is null

salvatore-campagna committed:
test: include a few more aggregation tests

ruflin commented on docs/reference/mapping/fields/ignored-field.asciidoc:
```diff
@@ -43,3 +43,20 @@ GET _search
 }
 --------------------------------------------------
+
+Since 8.13.0, the `_ignored` field supports aggregations as well.
```
I guess that is 8.14.0?

javanna replied:
yep

salvatore-campagna committed:
fix: remove existsQuery from legacy ignored field type

salvatore-campagna committed:
refactor: use ValueFetcher in innerGetFetch

salvatore-campagna committed:
Revert "refactor: use ValueFetcher in innerGetFetch"

This reverts commit 4a233df694c10eb328b1e1fe9f74cdff9b30731b.

salvatore-campagna committed:
fix: ignore term vectors generation for the _ignored field

javanna commented on server/src/main/java/org/elasticsearch/index/get/ShardGetService.java:
```diff
@@ -298,77 +293,51 @@ private GetResult innerGetFetch(
         if (mapperService.mappingLookup().isSourceEnabled() && fetchSourceContext.fetchSource()) {
+            Source source = loader.leaf(docIdAndVersion.reader, new int[] { docIdAndVersion.docId })
+                .source(leafStoredFieldLoader, docIdAndVersion.docId);
+
```
are these changes to ShardGetService needed?

salvatore-campagna committed:
docs: update version

salvatore-campagna committed:
fix: return _size only if required

salvatore-campagna committed:
test: update skip version

salvatore-campagna committed:
Merge branch 'main' into ignore_field_aggs

salvatore-campagna committed:
fix: include a mixed cluster test

salvatore-campagna committed:
fix: increase health check timeout

salvatore-campagna committed:
fix: expect two documents to be indexed

salvatore-campagna committed:
fix: yaml test skip version and additional phase

salvatore-campagna committed:
fix: phase name in yaml test

salvatore-campagna committed:
fix: include logic to deal with backward compatibility

salvatore-campagna committed:
refactor: mixed cluster test with exists and aggregation

salvatore-campagna committed:
comment: move it to the right if branch

salvatore-campagna committed:
fix: null pointer exception

salvatore-campagna committed:
fix: icnlude index settings and created version in mock

salvatore-campagna committed:
fix: icnlude index settings and created version in mock

salvatore-campagna committed:
fix: 8.14 bwc test

salvatore-campagna committed:
fix: incorrect index usage

salvatore-campagna committed:
fix: mixed cluster test check on old index version

salvatore-campagna committed:
fix: restore value of DUMP_OUTPUT_ON_FAILURE_PROP_NAME

salvatore-campagna committed:
fix: ESSingleNodeTestCase instead of ESIntegTestCase

salvatore-campagna committed:
fix: move test from IgnoreFieldIT to IgnoredFieldTypeTests

salvatore-campagna committed:
test: ignore field type selection logic

salvatore-campagna committed:
fix: use correct type depending on version

salvatore-campagna committed:
fix null pointer exception

salvatore-campagna committed:
fix: use parent exists query implementation

salvatore-campagna committed:
fix: remove incorrect null check

salvatore-campagna committed:
fix: read _ignored field from stored field for indices created before 8.14

salvatore-campagna committed:
fix: simplify bwc compatibility

salvatore-campagna committed:
fix: exists query and metadata stored field condition

salvatore-campagna commented:
@elasticsearchmachine run elasticsearch-ci/part-1 please

salvatore-campagna committed:
fix: do not skip metadata fields

salvatore-campagna committed:
nit: avoid using a temporary variable

salvatore-campagna committed:
fix: exclude meta fields when stored_fields = none

salvatore-campagna committed:
fix: exclude metadata fields handled by MetadataFetcher

salvatore-campagna commented:
@elasticsearchmachine test this please

salvatore-campagna commented:
buildkite test this

salvatore-campagna commented:
I am stuck here...after spending 3 days on this I can't figure out a way to make this work fixing bwc tests...whenever I fix something I break something else.

dnhatn commented:
@salvatore-campagna I can look into the BWC tests.

salvatore-campagna committed:
fix: undo fetch of stored fields

salvatore-campagna committed:
test: _ignored field upgrade test

salvatore-campagna commented:
I think the issue might be with old indices created by versions before 8.14. Those indices still have the _ignored field as stored and we might need to keep storing it as stored field to make sure that given an index the _ignored field has only stored fields (before 8.14) or doc value fields (from 8.14 on).

salvatore-campagna committed:
fix: include the _ignored stored field for old indices

salvatore-campagna committed:
Merge branch 'main' into ignore_field_aggs

javanna commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java:
```diff
@@ -14,30 +14,41 @@
+        final FetchFieldsContext fetchFieldsContext = fetchContext.fetchFieldsContext();
+        final StoredFieldsContext storedFieldsContext = fetchContext.storedFieldsContext();
+        final List<FieldAndFormat> fieldAndFormatList = fetchFieldsContext == null ? Collections.emptyList() : fetchFieldsContext.fields();
+        final FieldFetcher fieldFetcher = FieldFetcher.create(fetchContext.getSearchExecutionContext(), fieldAndFormatList);
```
I think that you don't want to call FieldFetcher if you have no fields to fetch? The conditional that you have removed above. I do see how you want to do the rest here as this phase needs now to also fetch metadata fields.

javanna replied:
Looking deeper, I'd suggest making this change separately: moving the fetch of metadata fields out from StoredFieldsPhase

salvatore-campagna replied:
Basically I moved fetching metadata fields here no matter if they are stored or use doc values. I understand that this can be moved to another phase, let's say `FecthMetadataFields`. ANyway from a functional point of view I would like to focus first on getting bwc work. Ideally once those are ok extracting this into another phase should not be an issue...

salvatore-campagna committed:
fix: address review comments

salvatore-campagna commented on server/src/main/java/org/elasticsearch/index/mapper/IgnoredFieldMapper.java:
```diff
@@ -46,29 +63,73 @@ public String typeName() {
     public void postParse(DocumentParserContext context) {
-        for (String field : context.getIgnoredFields()) {
-            context.doc().add(new StringField(NAME, field, Field.Store.YES));
+        if (context.indexSettings().getIndexVersionCreated().onOrAfter(IndexVersions.DOC_VALUES_FOR_IGNORED_META_FIELD)) {
```
@javanna I added this logic here to deal with the fact that for an index created by a version before 8.14 (the one where we introduce the change) we would like to keep writing stored values so that, given an index, we have _ignored being stored for already existing indices and doc values only for new ones. Does it make sense?

javanna replied:
Yes this bit makes sense to me, in that it's not field type dependent and covers for the indexing path.

salvatore-campagna commented:
buildkite test this

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/IgnoredFieldMapper.java:
```diff
@@ -46,29 +63,73 @@ public String typeName() {
+                    + typeName()
+                    + "' field are supported for indices created by version "
+                    + IndexVersions.DOC_VALUES_FOR_IGNORED_META_FIELD
+                    + " or higher"
```
I think we may want to avoid exposing version info in this exception. Perhaps the standard error for this would be fine, maybe there is no need to override this method?

salvatore-campagna replied:
Note that the standard error prints something that you commented on before like

```
"Fielddata is not supported on field [_ignored] of type [_ignored]"
```

salvatore-campagna committed:
refactor: metadata fields fetching logic and test

salvatore-campagna committed:
fix: IgnoredFieldMixedClusterUpgradeIT

salvatore-campagna committed:
Merge branch 'main' into ignore_field_aggs

salvatore-campagna committed:
fix: restore original after merging #106325

salvatore-campagna committed:
fix: restore original test outcome

salvatore-campagna committed:
fix: makre sure we get both old and new documents including _ignored fields

javanna commented on qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/upgrades/IgnoredMetaFieldAggregationIT.java:
```diff
@@ -0,0 +1,240 @@
+            if (getOldClusterIndexVersion().before(IndexVersions.DOC_VALUES_FOR_IGNORED_META_FIELD)) {
+                assertAggregateIgnoredMetadataFieldException(
+                    aggFirstIndex,
+                    "Fielddata is not supported on field [_ignored] of type [_ignored]"
```
The type is off in the message?

javanna replied:
The content type is kind of unexpected, as we set it to `_ignored`. But not a huge deal, I was not aware, lets' leave it as-is.

javanna reviewed (commented):
left some comments, minor things, we should be there.

salvatore-campagna committed:
fix: execute rolling upgrade test only if _ignored is stored in old version

salvatore-campagna committed:
fix: update skip version as of removing stored field for _ignored

javanna commented on qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/upgrades/IgnoredFieldRollingUpgradeIT.java:
```diff
@@ -67,10 +73,34 @@ public void testIgnoredMetaFieldExistsQuery() throws IOException {
+            // NOTE: from version 8.15 _ignored is not returned anymore by default when using the get api since
+            // `get` only allows fetching stored fields via `stored_fields` and _ignored is not stored anymore.
+            assertNull(doc.get(IgnoredFieldMapper.NAME));
+        }
```
Can you expand on this test? The get call below does not ask for the `_ignored` field explicitly. Am I getting confused or get did not return `_ignored` by default? 

Also, if you handle the two versions in the test, why execute it only against a specific version? That looks like something odd is happening?

salvatore-campagna replied:
Yes `GET` used to return the `_ignored` field up to version `8.4.3` while it is no longer the case from version `8.5.0`

RCQE commented:
## RCQE-bot: Commit Quality Assessment  🔎 
I've carefully evaluated the recent commits, and here's the current state of the commit quality: 
### Metrics 📊  
- Previous Commits: diff hunk in file [server/src/main/java/org/elasticsearch/index/mapper/IgnoredFieldMapper.java](https://github.com/elastic/elasticsearch/pull/101373/files/ee51ec899f9a158d1bc310732f3addafd3761e91#diff-1a208731fe86082d83ea058fc62187577478a6c1ee0cd3cbcd91d1b3b21ab8f4) in commit [c2232c97f0b5bf3e429e5e56136c018901c01fda](https://github.com/elastic/elasticsearch/pull/101373/commits/c2232c97f0b5bf3e429e5e56136c018901c01fda)
- Code Review: I think we may want to avoid exposing version info in this exception. Perhaps the standard error for this would be fine, maybe there is no need to override this method?
### Suggestions 💡 
- Need Review Again 🚫 : The commits do not currently meet the quality standards. I recommend not merging them into the main branch until the above suggestions have been implemented and a subsequent review confirms the quality. Please address the points mentioned and submit the commits for another round of review.

salvatore-campagna committed:
fix: skip version and get _ignored

salvatore-campagna committed:
fix: review comments

salvatore-campagna committed:
fix: update rolling upgrade tests

salvatore-campagna committed:
fix: update skip version after switching to 8.15

salvatore-campagna committed:
fix: undo dec ref changes

salvatore-campagna committed:
fix: review comments

salvatore-campagna committed:
Merge branch 'main' into ignore_field_aggs

salvatore-campagna committed:
fix: use public version string

salvatore-campagna committed:
fix: remove test for term vectors

salvatore-campagna committed:
docs: update version

salvatore-campagna committed:
fix: update version in reason

salvatore-campagna committed:
fix: remove tests included in another class

salvatore-campagna committed:
fix: error message and _ignored debug stored field

salvatore-campagna committed:
fix: remove _ignored

salvatore-campagna committed:
fix: remove SuppressWarnings

salvatore-campagna committed:
fix: remove _ignored

salvatore-campagna committed:
fix: error message in bwc test

salvatore-campagna committed:
fix: fetch _ignored field from doc values

salvatore-campagna commented on docs/reference/search/profile.asciidoc:
```diff
@@ -194,7 +194,7 @@ The API returns the following result:
             "load_source_count": 5
           },
           "debug": {
-            "stored_fields": ["_id", "_ignored", "_routing", "_source"]
```
This is not rerturned anymore because the field is not stored anymore.

javanna replied:
that makes sense, I was expecting this change.

salvatore-campagna commented on server/src/internalClusterTest/java/org/elasticsearch/search/source/MetadataFetchingIT.java:
```diff
@@ -123,12 +123,12 @@ public void testWithIgnored() {
         {
             GetResponse getResponse = client().prepareGet("test", "1").get();
             assertTrue(getResponse.isExists());
-            assertThat(getResponse.getField("_ignored"), nullValue());
```
This change is not required...will restore it.

salvatore-campagna committed:
fix: undo unecessary change

salvatore-campagna committed:
fix: non-null ingored fields after implementing fetch from doc values

salvatore-campagna committed:
Merge branch 'main' into ignore_field_aggs

javanna commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/get/120_stored_fields_ignored.yml:
```diff
@@ -1,9 +1,8 @@
-      reason: "_ignored metadata field uses doc values instead of stored field from version 8.14"
-      version: " - 8.13.99"
+      reason: "_ignored metadata field uses doc values instead of stored field from version 8.15"
+      version: " - 8.14.99"
```
Do we still need to skip this test if we implement bw comp?

salvatore-campagna replied:
I see this comment refers to outdated code...I remember I removed it...

javanna commented on server/src/main/java/org/elasticsearch/index/get/ShardGetService.java:
```diff
@@ -333,6 +337,19 @@ private GetResult innerGetFetch(
+                }
+                metadataFields.put(IgnoredFieldMapper.NAME, ignoredDocumentField);
+            }
+        }
```
Should we also make sure that we don't try to load `_ignored` from stored fields as well? I guess it's either loaded from doc values or from stored fields, but not from both?

javanna replied:
++

javanna commented on server/src/main/java/org/elasticsearch/index/get/ShardGetService.java:
```diff
@@ -333,6 +337,19 @@ private GetResult innerGetFetch(
+        // NOTE: when _ignored is requested via `stored_fields` we need to load it from doc values instead of loading it from stored fields.
+        // The _ignored field used to be stored, but as a result of supporting aggregations on it, it moved from using a stored field to
+        // using doc values.
+        if (storedFields != null && Arrays.asList(storedFields).contains(IgnoredFieldMapper.NAME)) {
```
Don't we need to execute the new logic based on index created version? What happens currently is that we try to load it from both doc values and stored fields regardless of index created version? That works but may not be the best choice? I don't know though what performance impact this has given that it's all done against a memory index.

salvatore-campagna replied:
Yes right, I did that later and I forgot to push actually.

javanna commented on server/src/internalClusterTest/java/org/elasticsearch/action/termvectors/GetTermVectorsIT.java:
```diff
@@ -1040,19 +1038,6 @@ public void testArtificialDocWithPreference() throws InterruptedException, IOExc
-        TermVectorsResponse resp = client().prepareTermVectors("index", "1").setSelectedFields("*").get();
-        assertThat(resp.getFields().terms("_ignored").size(), greaterThan(0L));
-    }
-
```
I think that given the changes you made to get, this can be restored. It should work again.

javanna reviewed (commented):
I left a couple of comments, but I think this is very close.

salvatore-campagna committed:
Revert "fix: remove test for term vectors"

This reverts commit ea20632fed7021ae14aed96158aa1adc144eb9cf.

salvatore-campagna committed:
fix: load _ignored from stored field or from doc values

salvatore-campagna committed:
fix: re-enable terms vector test

salvatore-campagna committed:
Merge branch 'main' into ignore_field_aggs

salvatore-campagna commented:
Ideally we should stop using `skip` in YAML tests for things that require versions above `8.14`. Anyway I see there are more yaml tests still using `skip` and I would prefer to handle those in a separate PR so that we apply changes to all yaml tests where that is required and avoid adding more stuff to this PR.

javanna commented on modules/parent-join/src/yamlRestTest/resources/rest-api-spec/test/30_inner_hits.yml:
```diff
@@ -140,7 +140,7 @@ profile fetch:
   - gt: { profile.shards.0.fetch.breakdown.load_stored_fields_count: 0 }
   - gt: { profile.shards.0.fetch.breakdown.load_stored_fields: 0 }
-  - match: { profile.shards.0.fetch.debug.stored_fields: [_id, _ignored, _routing, _source] }
+  - match: { profile.shards.0.fetch.debug.stored_fields: [_id, _routing, _source] }
```
I wonder if with this change, the skip above needs updating? Isn't it surprising that this test runs? 8.14 returns the `_ignored` field while 8.15 does not?

salvatore-campagna replied:
I fixed this but the other was ok.

javanna commented on qa/rolling-upgrade/src/javaRestTest/java/org/elasticsearch/upgrades/IgnoredMetaFieldRollingUpgradeIT.java:
```diff
@@ -0,0 +1,210 @@
+import java.util.Locale;
+import java.util.Map;
+
+public class IgnoredMetaFieldRollingUpgradeIT extends ParameterizedRollingUpgradeTestCase {
```
This is a great test to have!

javanna commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/370_profile.yml:
```diff
@@ -139,7 +139,7 @@ fetch nested source:
   - gt: { profile.shards.0.fetch.breakdown.load_stored_fields_count: 0 }
   - gt: { profile.shards.0.fetch.breakdown.load_stored_fields: 0 }
-  - match: { profile.shards.0.fetch.debug.stored_fields: [_id, _ignored, _routing, _source] }
+  - match: { profile.shards.0.fetch.debug.stored_fields: [_id, _routing, _source] }
```
Similar to above, I wonder if we need to update the skip above, that's what I would expect.

javanna replied:
I think I forgot about this commit that already made that change: https://github.com/elastic/elasticsearch/commit/19db4903d10689b736da8a60d96530424c1978d7

javanna commented on server/src/internalClusterTest/java/org/elasticsearch/index/mapper/IgnoredMetadataFieldIT.java:
```diff
@@ -0,0 +1,164 @@
+        }
+    }
+
+    public void testIgnoredMetadataFieldFetch() {
```
This is a bit of a duplicate of MetadatFetchingIT#testWithIgnored ? Is it needed?

salvatore-campagna replied:
I missed this comment...I can remove the test in another PR.

javanna reviewed (approved):
I left a couple questions, mostly around testing. LGTM otherwise! Great work!

salvatore-campagna committed:
fix: update skip version

salvatore-campagna committed:
Merge branch 'main' into ignore_field_aggs

salvatore-campagna committed:
Merge branch 'main' into ignore_field_aggs

salvatore-campagna commented:
Enables aggregations on the `_ignored` metadata field replacing the stored field
with doc values.

thecoop commented on server/src/main/java/org/elasticsearch/index/IndexVersions.java:
```diff
@@ -104,6 +104,7 @@ private static IndexVersion def(int id, Version luceneVersion) {
     public static final IndexVersion UPGRADE_TO_LUCENE_9_10 = def(8_503_00_0, Version.LUCENE_9_10_0);
     public static final IndexVersion TIME_SERIES_ROUTING_HASH_IN_ID = def(8_504_00_0, Version.LUCENE_9_10_0);
     public static final IndexVersion DEFAULT_DENSE_VECTOR_TO_INT8_HNSW = def(8_505_00_0, Version.LUCENE_9_10_0);
+    public static final IndexVersion DOC_VALUES_FOR_IGNORED_META_FIELD = def(8_505_00_1, Version.LUCENE_9_10_0);
```
@eyalkoren I've just spotted this - this should have incremented the NNN version of the version id, not the patch version. Check the comment right below the version constants for more info.

javanna replied:
@salvatore-campagna is the one two brought this to completion, thanks @thecoop for raising it.
